### ISSUE TITLE
PCL_EXPORTS seems to be implemented wrong

### ISSUE THREAD

**Issue by SunBlack** [PointCloudLibrary/pcl#2708]
I currently still trying to compile everything with MSVC. I could fix already a lot of issues, but there is an error I couldn't fix:
```
LNK2001	unresolved external symbol "public: static struct QMetaObject const pcl::cloud_composer::PropertiesModel::staticMetaObject" (?staticMetaObject@PropertiesModel@cloud_composer@pcl@@2UQMetaObject@@B)	pcl_cc_tool_fpfh_estimation
```
In general I could fix all linker errors, except where we export a target with Q_OBJECT:
```
    class PCL_EXPORTS AbstractTool : public QObject
    {
      Q_OBJECT
      //...
    }
```

If you search the web for this issue you always [find](https://stackoverflow.com/questions/19238693/export-qobject-based-class-to-dll), that we need `__declspec(dllimport)`. But `PCL_EXPORTS` expand always to `__declspec(dllexport)` or empty.

Further issue: `` PCL_ADD_LIBRARY` always add symbol `PCLAPI_EXPORTS`. This seems at first look correct, but I believe this is partially wrong.
* Case 1: Someone build PCL and use installed libraries. In this case everything is fine, because `PCLAPI_EXPORTS` is not defined
* Case 2: A PCL library (`LibA`) is using another PCL library (`LibB`). In this case `PCLAPI_EXPORTS` should be true for all header files part of `LibB` but no defined for `LibA` during building `LibB`. So I believe it would be correct if define `PCL_LibA_EXPORTS` and `PCL_LibB_EXPORTS` instead of just `PCL_EXPORTS`.

And we use should [CMake::GenerateExportHeader](https://cmake.org/cmake/help/v3.1/module/GenerateExportHeader.html).

**Comment by kunaltyagi** [comment]
Shall we revive the discussion here again @SunBlack ?

**Comment by SunBlack** [comment]
Yep, I still have this in mind, as I don't know maybe this could be the reason why some libs are so big under windows.

But in general I'm currently waiting for CMake changes as I think the best order would be:
- Reorder code (e.g. move tools out of module directories into tools directory): Wait for decision in #3837 therefore to prevent unnecessary git rename actions
- Add a macro like `PCL_ADD_MODULE` so all modules have a better CMake structure as it is currently really ugly with different  list for each type

After this cleanup we may start with issues like this, but currently it is to much copy & paste between all CMake files, even we could do it also now without CMake cleanup before.
